### ISSUE TITLE
Support arbitrary formatting helpers in FancyStrMixin

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6063]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6063|
|Type|enhancement|
|Created|2012-10-05 13:39:00Z|
|Branch|https://github.com/twisted/twisted/tree/fancyfancystrmixin-6063-2|

`FancyStrMixin` supports the printf-style string formatting specification for controlling how attributes are displayed.  This is handy, but printf-style only goes so far.  For example, on Python 3, there is no printf-style formatter which operates on bytes instance to produce a `"foo\x00"`-style output (`%s` comes with an unwanted `b` prefix).

Allowing a formatting function to be supplied would let applications format things any way they wish.  This is most directly needed by [#6057](https://github.com/twisted/twisted/issues/6057), but is probably more widely useful.

<details><summary>Searchable metadata</summary>

```
trac-id__6063 6063
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_fancyfancystrmixin_6063_2 branches/fancyfancystrmixin-6063-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349444340000000 1349444340000000
changetime__1349453925000000 1349453925000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [35958]) Branching to 'fancyfancystrmixin-6063'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35960]) Branching to 'fancyfancystrmixin-6063-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [35961]) Support a callable to format attribute values

refs [#6063](https://github.com/twisted/twisted/issues/6063)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

[Hope it works](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fancyfancystrmixin-6063-2)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Sigh. This is turning into an (even more) terrible API. I'm sure we can come up with a new, cleaner approach in our copious spare time. Meanwhile, this seems reasonable, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [35964]) Merge fancyfancystrmixin-6063-2

Author: exarkun
Reviewer: itamarst
Fixes: [#6063](https://github.com/twisted/twisted/issues/6063)

Add more flexibility to the formatting done by `twisted.python.util.FancyStrMixin` by
supporting arbitrary callables to format attribute values.
